data/blog/lost-in-the-haystack.md

```diff
@@ -0,0 +1,295 @@
+
+With this tool I can insert records into the database, run queries against those records, and evaluate the performance of those queries. I'm doing exactly that here to validate these schema changes. I've created a table called replays_test and I've added the `old` column which uses `Nullable(String)` and the `new` column which uses `LowCardinality(String)`. After that I bulk insert 1,000,000 rows and start playing around.
+
+We can run the following query to see the size of the columns on disk:
+
+```sql
+SELECT
+    name,
+    formatReadableSize(sum(data_compressed_bytes)) AS compressed_size,
+    formatReadableSize(sum(data_uncompressed_bytes)) AS uncompressed_size,
+    round(sum(data_uncompressed_bytes) / sum(data_compressed_bytes), 2) AS ratio
+FROM system.columns
+WHERE table = 'replays_test'
+GROUP BY name
+ORDER BY sum(data_compressed_bytes) DESC
+```
+
+Which outputs:
+
+```
+┌─name───────┬─compressed_size─┬─uncompressed_size─┬──ratio─┐
+│ old        │ 1.90 MiB        │ 4.47 MiB          │   2.36 │
+│ new        │ 370.76 KiB      │ 978.55 KiB        │   2.64 │
+└────────────┴─────────────────┴───────────────────┴────────┘
+```
+
+As you can see, our encodings have reduced our byte size by 80%. That's pretty cool and we didn't have to work too hard for it. We can also evaluate how these queries perform when reading these columns by running the following query:
+
+```
+SELECT
+    query_duration_ms,
+    read_rows,
+    read_bytes,
+    memory_usage,
+    query
+FROM system.query_log
+ORDER BY event_time DESC
+LIMIT 10
+```
+
+Which produces this output:
+
+```
+┌─query_duration_ms─┬─read_rows─┬─read_bytes─┬─memory_usage─┬─query────────────────┐
+│                24 │    245760 │    7047494 │     14303428 │ WHERE old = 'Safari' │
+│                14 │    286720 │    4874860 │      8743218 │ WHERE new = 'Safari' │
+└───────────────────┴───────────┴────────────┴──────────────┴──────────────────────┘
+```
+
+A pre-requisite to running this query is that you've run _other_ queries against your test table. But once you've done that you should see that our new, encoded column has significantly better performance than the old column. We read more rows with less memory usage and it takes us significantly less time to complete the query. Neat!
+
+Be careful running this migration. It can be tough depending on the size of your dataset.
+
+Side bar. Dropping nullability has implications for your query. For example, the `any` function will ignore nulls when computing its result. If you have an empty string value in your dataset that was previously marked as null your query may return that value rather than the non-empty string it previously returned. There's a simple solution to this. ClickHouse functions accept an `If` combinator and this combinator can be used to strip the empty values. Like so: `anyIf(column, notEmpty(column))`. The second argument to the function is any expression which returns a boolean result.
```

Comment:
```suggestion
Sidebar. Dropping nullability has implications for your query. For example, the `any` function will ignore nulls when computing its result. If you have an empty string value in your dataset that was previously marked as null your query may return that value rather than the non-empty string it previously returned. There's a simple solution to this. ClickHouse functions accept an `If` combinator and this combinator can be used to strip the empty values. Like so: `anyIf(column, notEmpty(column))`. The second argument to the function is any expression which returns a boolean result.
```